engine/src/main/java/org/terasology/physics/engine/VoxelFluidWorldSystem.java

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.physics.engine;
+
+import com.badlogic.gdx.physics.bullet.collision.VoxelCollisionAlgorithmWrapper;
+import com.badlogic.gdx.physics.bullet.collision.btCollisionObject;
+import com.badlogic.gdx.physics.bullet.collision.btCollisionShape;
+import com.badlogic.gdx.physics.bullet.collision.btVoxelInfo;
+import com.badlogic.gdx.physics.bullet.collision.btVoxelShape;
+import com.badlogic.gdx.physics.bullet.dynamics.btDiscreteDynamicsWorld;
+import com.badlogic.gdx.physics.bullet.dynamics.btRigidBody;
+import com.badlogic.gdx.physics.bullet.linearmath.btDefaultMotionState;
+import gnu.trove.set.hash.TShortHashSet;
+import org.joml.Matrix4f;
+import org.joml.Vector3f;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.physics.StandardCollisionGroup;
+import org.terasology.physics.bullet.BulletPhysics;
+import org.terasology.physics.bullet.shapes.BulletCollisionShape;
+import org.terasology.registry.In;
+import org.terasology.world.OnChangedBlock;
+import org.terasology.world.WorldComponent;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockComponent;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.ChunkProvider;
+import org.terasology.world.chunks.event.BeforeChunkUnload;
+import org.terasology.world.chunks.event.OnChunkLoaded;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.terasology.physics.bullet.BulletPhysics.AABB_SIZE;
+
+/**
+ * Manages voxel shape for fluid and updates collision state between Bullet and Terasology
+ */
+@RegisterSystem
+public class VoxelFluidWorldSystem extends BaseComponentSystem {
```

Comment:
I've been trying to maintain a consistent distinction between "fluid", the substances registered with the FluidRegistry, which can fill containers and pipes, and "liquid", blocks which have the `liquid` attribute. This is liquid, not fluid.